### PR TITLE
ValidatingPublisher

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by simonbasle** [comment]
@smaldini renamed `ValidatingPublisher`, I think that is a good first MVP, ready for review

File: reactor-test/src/main/java/reactor/test/publisher/ValidatingPublisher.java [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+
+	/**
+	 * Create a {@link Misbehavior misbehaving} {@link ValidatingPublisher}.
+	 *
+	 * @param first the first allowed {@link Misbehavior}
+	 * @param rest additional optional misbehaviors
+	 * @param <T> the type of the publisher
+	 * @return the new misbehaving {@link ValidatingPublisher}
+	 */
+	public static <T> ValidatingPublisher<T> createMisbehaving(Misbehavior first, Misbehavior... rest) {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ValidatingPublisher.java]: `Misbehaving` name isn't perfect, maybe find another name

File: reactor-test/src/main/java/reactor/test/publisher/ValidatingPublisher.java [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+	 * @return the new {@link ValidatingPublisher}
+	 */
+	public static <T> ValidatingPublisher<T> create() {
+		return new ValidatingPublisher<T>();
+	}
+
+	/**
+	 * Possible misbehavior for a {@link ValidatingPublisher}.
+	 */
+	public enum Misbehavior {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/publisher/ValidatingPublisher.java]: same here, should the name be changed? also, is the enum enough? a builder could be used as an alternative. This boils down to "do we expect users to often allow a lot of misbehaviors?" in which case enums would maybe be less convenient to use.

**Comment by simonbasle** [comment]
as per @rstoyanchev feedback, renamed to `TestPublisher`. this is now an interface, and `expectXXX` methods have been renamed back to `assertXXX`. The only remaining questions are:

- [ ] is the enum expressive and simple enough for what we want (a rather coarse-grained set of misbehaviors that the publisher usually guard against but can overlook)?
- [ ] does the name `Misbehavior`/`misbehaving` convey the right idea or should the logic of the naming be inverted somehow? (eg. `Misbehavior.ALLOW_NULLS` and `createMisbehaving(ALLOW_NULLS)` vs `BehaviorCheck.REJECT_NULLS` and `createExcludingChecks(REJECT_NULLS`)

**Comment by rstoyanchev** [comment]
Perhaps `noncompliant` and `violation` as in "spec violation" ...

**Pull request merged**